modules/relayers/src/lib.rs

```diff
@@ -102,9 +109,194 @@ pub mod pallet {
+					None => fail!(Error::<T>::NotRegistered),
+				};
+
+				// we can't deregister until `valid_till + 1`
+				ensure!(
+					registration.valid_till < frame_system::Pallet::<T>::block_number(),
+					Error::<T>::RegistrationIsStillActive,
+				);
+
+				// if stake is non-zero, we should do unreserve
+				if !registration.stake.is_zero() {
+					Self::do_unreserve(&relayer, registration.stake)?;
+				}
+
+				Self::deposit_event(Event::<T>::Deregistered { relayer: relayer.clone() });
+
+				*maybe_registration = None;
+
+				Ok(())
+			})
+		}
 	}
 
 	impl<T: Config> Pallet<T> {
+		/// Returns true if given relayer registration is active at current block.
+		///
+		/// This call respects both `RequiredStake` and `RequiredRegistrationLease`, meaning that
+		/// it'll return false if registered stake is lower than required or if remaining lease
+		/// is less than `RequiredRegistrationLease`.
+		pub fn is_registration_active(relayer: &T::AccountId) -> bool {
+			let registration = match Self::registered_relayer(relayer) {
+				Some(registration) => registration,
+				None => return false,
+			};
+
+			// registration is inactive if relayer stake is less than required
+			if registration.stake < Self::required_stake() {
+				return false
+			}
+
+			// registration is inactive if it ends soon
+			let remaining_lease = registration
+				.valid_till
+				.saturating_sub(frame_system::Pallet::<T>::block_number());
+			if remaining_lease <= Self::required_registration_lease() {
+				return false
+			}
+
+			true
+		}
+
+		/// Slash and `deregister` relayer. This function slashes all staked balance.
+		///
+		/// It may fail inside, but error is swallowed and we only log it.
+		pub fn slash_and_deregister(
+			relayer: &T::AccountId,
+			slash_destination: RewardsAccountParams,
+		) {
+			let registration = match RegisteredRelayers::<T>::take(relayer) {
+				Some(registration) => registration,
+				None => {
+					log::trace!(
+						target: crate::LOG_TARGET,
+						"Cannot slash unregistered relayer {:?}",
+						relayer,
+					);
+
+					return
+				},
+			};
+
+			match T::StakeAndSlash::repatriate_reserved(
+				relayer,
+				slash_destination,
+				registration.stake,
+			) {
+				Ok(failed_to_slash) if failed_to_slash.is_zero() => {
+					log::trace!(
+						target: crate::LOG_TARGET,
+						"Relayer account {:?} has been slashed for {:?}. Funds were de;posited to {:?}",
+						relayer,
+						registration.stake,
+						slash_destination,
+					);
+				},
+				Ok(failed_to_slash) => {
+					log::trace!(
+						target: crate::LOG_TARGET,
+						"Relayer account {:?} has been partially slashed for {:?}. Funds were deposited to {:?}. \
+						Failed to slash: {:?}",
+						relayer,
+						registration.stake,
+						slash_destination,
+						failed_to_slash,
+					);
+				},
+				Err(e) => {
+					// TODO: document this. Where?
+
+					// it may fail if there's no beneficiary account. For us it means that this
```

Comment:
Afaik we can't just burn tokens, as this changes the total issuance.

> Otherwise won't the malicious relayer be able to deregister() and get the amount back ?

Yes, he can. But there's not much we can do - that's a misconfiguration.